walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,170 @@
 
 package com.google.sps;
 
+import java.lang.Math;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  /**
+   * Determine available times for a meeting to take place given a list of Events and a Meeting Request.
+   * @param events the list of events to take into consideration when finding available times
+   * @param request the meeting request with meeting attendees and duration
+   * @return A list of available TimeRanges
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Meeting cannot last longer than 24 hours
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) return Arrays.asList();
+
+    Collection<TimeRange> availableTimesWithOptionalAttendees = queryHelper(events, request, true);
+
+    // If there are no available times with all (mandatory and optional attendees) then only consider mandatory 
+    if(availableTimesWithOptionalAttendees.size() == 0 && request.getAttendees().size() > 0) {
+      return queryHelper(events, request, false);
+    }
+
+    return availableTimesWithOptionalAttendees;
+  }
+
+  /**
+   * Helper method for the query method which finds the available times to meet given a list of events and a MeetingRequest,
+   * depending on whether optional attendees are considered or not.
+   * @param events the list of events to take into consideration when finding available times
+   * @param request the meeting request with the meeting attendees and duration
+   * @param considerOptional whether optional attendees are accounted for or not
+   * @return available times for the group to meet
+   *
+   */
+  private Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request, boolean considerOptional) {
+    Collection<String> attendees = new HashSet<String>(request.getAttendees());
+    if(considerOptional) {
+      attendees.addAll(request.getOptionalAttendees());
+    }
+    
+    // Needs to have at least 1 attendee at a meeting
+    if(attendees.size() == 0) return Arrays.asList(TimeRange.WHOLE_DAY);
+
+    // For a time to be unavailable, there must be someone at the event who is in the meeting attendees list
+    List<TimeRange> unavailableTimes = new ArrayList(convertAttendeesAndEventsToUnavailableTimes(events, attendees));
```

Comment:
I really like the way you've structured your `queryHelper` function - having helpers to do each of the sub-tasks here makes for clear, readable code that's easy to follow!

Reply:
Thanks! There's some logic in places that doesn't make clear sense without comments so I wanted to break it down